README.md
# EdgeFlush

<!--
[![Latest Version on Packagist](https://img.shields.io/packagist/v/area17/edge-flush.svg?style=flat-square)](https://packagist.org/packages/area17/edge-flush)
-->
[![GitHub PHPUnit Action Status](https://img.shields.io/github/workflow/status/area17/edge-flush/phpunit?label=PHPUnit)](https://github.com/area17/edge-flush/actions?query=workflow%3Aphpunit+branch%3A1.x)
[![GitHub PHPStan Action Status](https://img.shields.io/github/workflow/status/area17/edge-flush/phpstan?label=PHPStan)](https://github.com/area17/edge-flush/actions?query=workflow%phpstan+branch%3A1.x)
<!--
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/area17/edge-flush/Check%20&%20fix%20styling?label=code%20style)](https://github.com/area17/edge-flush/actions?query=workflow%3A"Check+%26+fix+styling"+branch%3A1.x)
[![Total Downloads](https://img.shields.io/packagist/dt/area17/edge-flush.svg?style=flat-square)](https://packagist.org/packages/area17/edge-flush)
-->

EdgeFlush is Laravel package intended to help developers manage CDN granular caching and invalidations. Having Akamai, CloudFront (or any other CDN) in front of a website, data modification usually forces us to bust the whole cache, leading to a website slow (for the first users) until the whole cache is rebuilt, and if the "first user" is Google Bot, for example, this can also impact on your website's rank. This pacakge aims to do invalidations granularly.

## Feature list

- Granular invalidation: this package will create a collection of all models that impacts one page and when one of those models change, all pages that had that model rendered in previous requests will be purged from CDN.
- Granular control of Cache-Control headers: you will be able to configure it differently per request, telling the CDN to store some pages for one week and others for 5 seconds, for example.
- Single [Akamai Edge Cache Tag](#akamai-edge-cache-tags) relating to all models touched by a page render.
- Define different strategies for Cache-Control: web pages may have a different cache strategy than api endpoints.
- Prevents from caching pages containing forms.
- Caches only frontend pages, leaving the CMS uncashed, if needed.
- [Re-warm](#rewarming-cache) pages purged from cache.
- Strip cookies from cachable responses.
- Disable caching for some pages using a middlware.
- Define HTTP methods that allow caching or not: cache GET but not POST.
- Define HTTP response status codes that allow caching or not: Cache 200 and 301 but not 400+ status codes.
- Define what routes can and cannot be cached by CDN.
- Define what type of responses can be cached: cache Response but not JsonResponse, for example.
- Define what Model classes can be cached or not.
- Remember what pages have been cached and command your CDN service to burst only those when you save something on your backend.
- Supports CloudFront invalidations.
- Supports Akamai EdgeCacheTags invalidations.
- Allow override of Services and easy implementation to support new CDN Services.
- [Spatie's Laravel Response Cache](#laravel-response-cache-integration) granular invalidations.

## Installation

Install the package via composer:

```bash
composer require area17/edge-flush
```

Publish the config file with:

```bash
php artisan vendor:publish --provider="A17\EdgeFlush\ServiceProvider"
```

And run the migrations:

```bash
php artisan migrate
```

## Dependencies

The supported CDN services have these package dependencies that you need to choose according to your setup:

Akamai: [akamai-open/edgegrid-auth](https://github.com/akamai/AkamaiOPEN-edgegrid-php)
CloudFront: [aws/aws-sdk-php](https://github.com/aws/aws-sdk-php)

## Usage

Do a full read on the `config/edge-flush.php` there's a lot of configuration items and we tried to document them all.

Define your CDN service class on `config/edge-flush.php`:

``` php
'classes' => [
    'cdn' => A17\EdgeFlush\Services\CloudFront\Service::class,
    
    ...
]
```

Add the trait `A17\EdgeFlush\Behaviours\CachedOnCDN` to your models and repositories.

Call `$this->invalidateCDNCache($model)` every time a model (on your base model or repository save() method). This example takes in consideration [Twill's](https://twill.io/) repositories:

``` php
public function afterSave($object, $fields)
{
    $this->invalidateCDNCache($object);

    parent::afterSave($object, $fields);
}
```

Call `$this->cacheModelOnCDN($model)` method on model's `getAttribute()`:

``` php
public function getAttribute($key)
{
    $this->cacheModelOnCDN($this);

    return parent::getAttribute($key);
}
```

Add the Middlware to the `Kernel.php` file:

```
protected $middleware = [
    \A17\EdgeFlush\Middleware::class,
    ...
];
```

Cache-Control max-age and s-maxage is set automatically, but if you need to change it depending on the current request you can use the following method:

``` php
CacheControl::setMaxAge(5000); // in seconds

CacheControl::setMaxAge('1 month'); // as a DateTime string period

CacheControl::setSMaxAge('2 weeks');
```

If you want to invalidate your paths in batches, add a scheduler setting the desired frequency for this to happen:

``` php
protected function schedule(Schedule $schedule)
{
    $schedule->job(new PurgeTags())->everyMinute();
}
```

You need to enable the package and the warmer on your `.env` file

``` sh
EDGE_FLUSH_ENABLED=true
EDGE_FLUSH_WARMER_ENABLED=true
```

## CDN third-party service configuration

Please check the respective environment variables needed for supported services to work:

- [Akamai](https://github.com/area17/edge-flush/blob/unstable/config/edge-flush.php#L188)
- [CloudFront](https://github.com/area17/edge-flush/blob/unstable/config/edge-flush.php#L195)

## Rewarming cache

Purged cache pages can load slowly for the next users or even Google Bot, if you want to prevent this you can enable (on config) the cache warmer and add the job to the schedule:

``` php
protected function schedule(Schedule $schedule)
{
    $schedule->job(new WarmCache())->everyMinute();
}
```

Note that the most hit (or frequently updated) pages will be warmed first.

## Akamai Edge Cache Tags

Akamai has a 128 bytes limit for the tag list, so if one page is impacted by lots of models, we would have no other way than busting the whole cache every time. This package creates a single Edge Cache Tag that relates to all models touched when the page was rendered, and adds it yo the response header:

```
edge-cache-tag: app-production-7e0ae085d699003a64e5fa7b75daae3d78ace842
```

## Invalidating the full cache from the command line

In case you need to invalidate the whole CDN cache locally or on a deployment routing, you can:

```
php artisan edge-flush:invalidate-all
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](.github/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

-   [AREA 17](https://github.com/area17)
-   [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `EdgeFlush` will be documented in this file.

## 1.4.3 - 2022-10-27
### Added
- Fix URLs not being correctly parsed with username and password

## 1.4.2 - 2022-10-26
### Added 
- Hook up update events to the Laravel Eloquent event system

## 1.4.0 - 2022-10-25
## 1.4.1 - 2022-10-25
### Changed
- Move up to PHPStan Level 9

## 1.3.14 - 2022-09-28
### Changed
- Improve Guzzle connection errors handling
- Allow invalidations when disabled
- Prevent CDN service instantiation until it's really needed
- Improve the way warmer reset tags and URLs

## 1.3.XX - 2022-10-XX
### Added 
- Full cache invalidation command: php artisan edge-flush:invalidate-all


## 1.3.10 - 2022-09-26
### Fixed
- Correctly setting was_purged_at update SQL

## 1.3.9 - 2022-09-26
### Changed
- Improve performance by not instantiating the CDN service until it's really needed

## 1.3.5 - 2022-09-14
### Changed
- Force invalidate all to clear all tags and urls

## 1.3.4 - 2022-09-13
### Changed
- Remove purged status from URLs that were hit

## 1.3.3 - 2022-09-13
### Changed
- Improved invalidation URLs query performance

## 1.3.2 - 2022-09-13
### Fixed
- Fix warmer not dispatching requests

## 1.3.1 - 2022-09-13
### Fixed
- Properly mark URLs as purged

## 1.3.0 - 2022-09-13
### Added
- Add a new Invalidation class to control invalidation items and the invalidation itself
### Changed
- Strong query and update optimization
### Fixed
- Fix all PHPStan level 8 errors

## 1.0.2 - 2021-11-10
### Fixed
- Disabled state generating recursion

## 1.0.1 - 2021-11-10
### Added
- Alow disabling on runtime
### Fixed
- EdgeFlush was not easily disablable 

## 1.0.0 - 2021-07-28
-   initial release


config/cdn.php
<?php

use Illuminate\Support\Str;
use Illuminate\Http\Response;
use A17\CDN\Support\Constants;
use Illuminate\Http\JsonResponse;

return [
    /**
     * Enable/disable the pacakge
     */
    'enabled' => env('CDN_ENABLED', false),

    /**
     * Configure here the default strategies used internally.
     * You can still manually set the current strategy at run-time.
     */
    'built-in-strategies' => [
        'cache' => 'dynamic',

        'micro-cache' => 'micro',

        'zero-cache' => 'zero',
    ],

    /**
     * Service classes
     *
     * Supported CDN services: Akamai, CloudFront
     *
     */
    'classes' => [
        'cdn' => A17\CDN\Services\Akamai\Service::class,

        'cache-control' => A17\CDN\Services\CacheControl::class,

        'tags' => A17\CDN\Services\Tags::class,

        'warmer' => A17\CDN\Services\Warmer::class,

        'response-cache' => A17\CDN\Services\ResponseCache\Service::class,
    ],

    /**
     * Caching strategies.
     *
     * Refer to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control
     *
     * The default "micro-cache" strategy takes in consideration that a 5 seconds cache
     * is better than NO-CACHE, and if your application gets hit by a DDoS attack
     * only 1 request every 5 seconds (per page) will hit your servers.

     * These are the supported directives:
     *
     *   public
     *   private
     *   no-cache
     *   no-store
     *   max-age=<seconds>
     *   s-maxage=<seconds>
     *   max-stale=<seconds> // not supported yet
     *   min-fresh=<seconds> // not supported yet
     *   stale-while-revalidate=<seconds> // not supported yet
     *   stale-if-error=<seconds> // not supported yet
     *   must-revalidate
     *   proxy-revalidate
     *   immutable
     *   no-transform
     *   only-if-cached
     *
     */
    'strategies' => [
        'dynamic' => ['max-age', 'public'], // built-in

        'micro' => ['max-age=5', 'public'], // built-in

        'zero' => ['max-age=0', 'no-store'], // built-in

        'api' => ['max-age=20', 'public', 'no-store'], // custom
    ],

    /**
     * Define how max-age will work.
     *
     * default: in case max-age is not configured at runtime, and
     * pages are set to be cached, this is the default value.
     *
     * strategy: defines how the setter will behave:
     *    min = will use the minumum value set
     *    last = will use the last value set
     *
     */
    'max-age' => [
        'default' => 1 * Constants::WEEK,

        'strategy' => 'min', // min, last
    ],

    /**
     * In case max-age is not configured at runtime, and
     * pages are set to be cached, this is the default value.
     */
    'cache-control' => ['max-age' => 1 * Constants::WEEK],

    /**
     * We suppose only your frontend application needs to be
     * behind a CDN, please use this closure to tell the service
     * if the request was requested by the frontend or not.
     *
     * To allow everything you can just set
     *
     *    'frontend-checker' => true,
     *
     */
    'frontend-checker' => A17\CDN\Services\FrontendChecker::class,

    /**
     * List of cache control headers to add to responses
     */
    'headers' => [
        'cache-control' => ['Cache-Control', 'X-Cache-Control'],

        'tags' => ['Edge-Cache-Tag', 'X-Cache-Tag'],
    ],

    /**
     * Usually pages that contains forms should not be cached. Here you can
     * define if you want this to checked and what are valid form strings for
     * your application.
     */
    'valid_forms' => [
        'enabled' => true,

        'strings' => [
            '<form',
            '<input type="hidden" name="_token" value="%CSRF_TOKEN%"',
        ],
    ],

    /**
     * Allowed responses
     */
    'responses' => [
        'cachable' => [
            Illuminate\Http\Response::class,
            Illuminate\Http\JsonResponse::class,
        ],

        'not-cachable' => [],
    ],

    /**
     * Allowed routes. You can also tell the package if you want to cache
     * routes that have no name.
     */
    'routes' => [
        'cachable' => [],

        'not-cachable' => ['*.ticket', 'newsletter*', 'api.*'],

        'cache_nameless_routes' => true,
    ],

    /**
     * Allowed URLs.
     */
    'urls' => [
        'cachable' => ['**/**'],

        'not-cachable' => ['**/debug/**'],
    ],

    /**
     * Allowed methods
     */
    'methods' => [
        'cachable' => ['GET'],

        'not-cachable' => [],
    ],

    /**
     * Allowed response statuses
     */
    'statuses' => [
        'cachable' => [200, 301],

        'not-cachable' => [],
    ],

    /**
     * Tags configuration. Here you can exclude model classes and
     * define the format of the page-related tags.
     */
    'tags' => [
        'excluded-model-classes' => [
            '\Models\Translations*',
            '\Models\Slugs*',
            '\Models\Revisions*',
        ],

        'format' => 'app-%environment%-%sha1%',
    ],

    /**
     * Invalidations.
     *
     * Invalidations can be sent one by one or in batch. In this section you can configure
     * this and also set the limits.
     *
     * single invalidations are executed immediately after a model is updated.
     * batch invalidations are executed every x minutes.
     *
     * batch.max_paths: what's is the limit before we invalidate the whole site?
     * batch.site_root: what's the site root path that should be invalidated?
     */
    'invalidations' => [
        'method' => 'invalidate', // invalidate, delete

        'type' => 'single', // single, batch

        'batch' => [
            'max_tags' => 2999, /// CloudFront limit is 3000
            'site_roots' => ['/*'],
        ],
    ],

    /**
     * Services configuration
     */
    'services' => [
        'akamai' => [
            'host' => env('CDN_AKAMAI_HOST'),
            'access_token' => env('CDN_AKAMAI_ACCESS_TOKEN'),
            'client_token' => env('CDN_AKAMAI_CLIENT_TOKEN'),
            'client_secret' => env('CDN_AKAMAI_CLIENT_SECRET'),
            'invalidate_all_paths' => ['*'],
        ],

        'cloud_front' => [
            'sdk_version' => env('CDN_CLOUD_FRONT_SDK_VERSION', '2016-01-13'),

            'region' => env(
                'CDN_AWS_DEFAULT_REGION',
                env('AWS_DEFAULT_REGION', 'us-east-1'),
            ),

            'distribution_id' => env('CDN_AWS_CLOUDFRONT_DISTRIBUTION_ID'),

            'key' => env('CDN_AWS_CLOUDFRONT_KEY', env('AWS_ACCESS_KEY_ID')),

            'secret' => env(
                'CDN_AWS_CLOUDFRONT_SECRET',
                env('AWS_SECRET_ACCESS_KEY'),
            ),

            'invalidate_all_paths' => ['/*'],
        ],
    ],

    /**
     * Purged cache can be rewarmed. Enable and configure it here.
     *
     * max_urls: how many urls max should the warmer try to warm per session?
     *
     * max_time: how much time the whole job warming session can take?
     *
     * connection_timeout: to warm up a page we don't actually need to get the page contents
     *                     that's why the warmer will kill the connection (possibly) before
     *                     the server is able to respond with contents. The idea here is
     *                     to speed up the warming process and save bandwidth.
     *
     * concurrent_requests: how many concurrent requests the warmer should dispatch per session?
     *
     * warm_all_on_purge: if the whole CDN cache is purged, do you wish to warm back all pages?
     *
     * wait_before_warming: invalidating tags can take time, maybe it's good to wait a couple
     *                      of minutes before warming purged urls.
     *
     */
    'warmer' => [
        'enabled' => env('CDN_WARMER_ENABLED', false),

        'max_urls' => 100,

        'max_time' => Constants::MILLISECOND * 750,

        'connection_timeout' => Constants::MILLISECOND * 15,

        'concurrent_requests' => 50,

        'warm_all_on_purge' => true,

        'wait_before_warming' => Constants::MINUTE * 2,
    ],

    /**
     * When a page is to be cached by the CDN this package will strip the following
     * cookies from all responses.
     */
    'strip_cookies' => [
        'XSRF-TOKEN',
        Str::slug(env('APP_NAME', 'laravel'), '_') . '*',
    ],
];
